sw/device/tests/crypto/cryptotest/firmware/ibex_sca.c

```diff
@@ -129,22 +705,142 @@ status_t handle_ibex_sca_register_file_write(ujson_t *uj) {
+  TRY(ujson_deserialize_ibex_sca_test_fvsr_t(uj, &uj_data));
+  if (uj_data.num_iterations >= 256) {
+    LOG_INFO("Error: num_iterations >= 256");
+    return ABORTED();
+  }
+
+  // Generate FvsR values.
+  uint32_t values[256];
+  bool sample_fixed = true;
+  for (int i = 0; i < uj_data.num_iterations; i++) {
+    if (sample_fixed) {
+      values[i] = uj_data.fixed_data;
+    } else {
+      values[i] = prng_rand_uint32();
+    }
+    sample_fixed = prng_rand_uint32() & 0x1;
+  }
+
+  for (int i = 0; i < uj_data.num_iterations; i++) {
+    // Initialize temporary registers with reference values.
+    asm volatile("mv x5, %0" : : "r"(0));
+    asm volatile("mv x6, %0" : : "r"(0));
+    asm volatile("mv x7, %0" : : "r"(0));
+    asm volatile("mv x28, %0" : : "r"(values[i]));
+    asm volatile("mv x29, %0" : : "r"(values[i]));
+    asm volatile("mv x30, %0" : : "r"(values[i]));
+    asm volatile(NOP100);
+    // SCA code target.
     sca_set_trigger_high();
-    // Write provided data into register file.
-    asm volatile("mv x5, %0" : : "r"(uj_data.data[0]));
-    asm volatile("mv x6, %0" : : "r"(uj_data.data[1]));
-    asm volatile("mv x7, %0" : : "r"(uj_data.data[2]));
-    asm volatile("mv x28, %0" : : "r"(uj_data.data[3]));
-    asm volatile("mv x29, %0" : : "r"(uj_data.data[4]));
-    asm volatile("mv x30, %0" : : "r"(uj_data.data[5]));
-    asm volatile("mv x31, %0" : : "r"(uj_data.data[6]));
+    // Give the trigger time to rise.
+    asm volatile(NOP100);
+    // Copy registers.
+    asm volatile("mv x5, x28");
+    asm volatile("mv x6, x29");
+    asm volatile("mv x7, x30");
     sca_set_trigger_low();
   }
-  // Acknowledge test.
+
+  // Write back last value written into the RF to validate generated data.
   ibex_sca_result_t uj_output;
-  uj_output.result = 0;
+  uj_output.result = values[uj_data.num_iterations - 1];
+  RESP_OK(ujson_serialize_ibex_sca_result_t, uj, &uj_output);
+  return OK_STATUS(0);
+}
+
+/**
+ * ibex.sca.register_file_read_batch_random
+ *
+ * This SCA penetration test executes the following instructions:
+ * - Generate num_iteration random values using the software LFSR.
+ * - Loop num_iterations:
+ *  - Set trigger
+ *  - Read data from RF
+ *  - Unset trigger
+ *
+ * SCA traces are captured during trigger_high & trigger_low.
+ *
+ * @param uj The received uJSON data.
+ */
+status_t handle_ibex_sca_register_file_read_batch_random(ujson_t *uj) {
+  // Get number of iterations.
+  ibex_sca_batch_t uj_data;
+  TRY(ujson_deserialize_ibex_sca_batch_t(uj, &uj_data));
+  if (uj_data.num_iterations >= 256) {
+    LOG_INFO("Error: num_iterations >= 256");
+    return ABORTED();
+  }
+
+  // Generate random values.
+  uint32_t values[256];
+  for (int i = 0; i < uj_data.num_iterations; i++) {
+    values[i] = prng_rand_uint32();
+  }
+
+  for (int i = 0; i < uj_data.num_iterations; i++) {
+    // Initialize temporary registers with reference values.
+    asm volatile("mv x5, %0" : : "r"(0));
```

Comment:
move this to a static function and share across the file? (DRY principle)

Reply:
You mean the generation of the random values? I've created a two static functions for generating random and FvsR values instead of repeating the same code.

Reply:
I mean this asm block that initializes temp registers with values

Reply:
sorry, inline function or macro? just to help keep everything in sync over time. I.e., if you change something, you have to change it everywhere, makes things hard to maintain. Unless there are SCA reasons you have like you stated below.

Reply:
Done, thanks :)